src/torrent.cpp

```diff
@@ -7429,23 +7429,51 @@ bool is_downloading_state(int const st)
 		if (lhs->on_parole() != rhs->on_parole())
 			return lhs->on_parole();
 
-		// prefer to disconnect peers that send data at a lower rate
+		// prefer to disconnect peers that chokes us
+		if (lhs->is_choked() != rhs->is_choked())
+			return lhs->is_choked();
+
+		// prefer to disconnect peers that send data at a lower rate over those waiting to download
+		// the assumption is that a connection waiting to start downloading has the same potential 
+		// to be faster than the current slowest downloading connection than a new connection that 
+		// replaces a waiting connection - and disconnecting a waiting connection starts a new wait
+		// however ... to avoid torrents which wait forever there needs to be a maximum wait time
+		// after which a waiting connection is disconnected
 		std::int64_t lhs_transferred = lhs->statistics().total_payload_download();
 		std::int64_t rhs_transferred = rhs->statistics().total_payload_download();
+		bool const is_waiting = lhs_transferred * rhs_transferred == 0;
 
+		// for this commit, the maximum time is a constant - but it could be replaced by a new 
+		// setting peer_turnover_maximum_wait in a future commit
+		std::int64_t const peer_turnover_maximum_wait = 60;
 		time_point const now = aux::time_now();
 		std::int64_t const lhs_time_connected = total_seconds(now - lhs->connected_time());
 		std::int64_t const rhs_time_connected = total_seconds(now - rhs->connected_time());
+		
+		// if one connection is waiting
+		if ((lhs_transferred != rhs_transferred) && (is_waiting))
+		{
+			// One of lhs_transferred & rhs_transferred is zero - get connected time for waiting connection
+			std::int64_t const waiting_time = 
+				((lhs_transferred * rhs_time_connected) + (rhs_transferred * lhs_time_connected)) 
+				/
+				(lhs_transferred + rhs_transferred);
+			if (waiting_time <= peer_turnover_maximum_wait)
+				return lhs_transferred != 0;
+			else return lhs_transferred == 0;
```

Comment:
I think this is a really involved way of saying what you want to say. But it's also not clear that it makes sense to do this, in general. A peer that has never sent anything to you, will be prioritized. Over time you would presumably accrue many such peers.

Reply:
In theory, you shouldn't accumulate many such peers waiting to download because over time they should start to actually download. But in reality:

a. Some peers will take a long time to start to download;
b. If a specially crafted malicious peer can connect and deliberately wait indefinitely, they could cause a DoS.  

And this is why I only prioritise disconnects of downloading connections over waiting connections if the wait time is < Peer_Turnover_Maximum_Wait - currently a constant, but it could be a setting. If a connection has waited longer than this then the waiting connection is prioritised for disconnection over the downloading connection, thus avoiding the accumulation of waiting connections.

Reply:
I think the other peer should be assumed to be malicious

Reply:
Sorry - I don't understand what you mean. This doesn't assume maliciousness as such, it just expects a waiting connection to start downloading in a reasonable timescale, and if it hasn't started downloading by Peer_Turnover_Maximum_Wait it prioritises the waiting connection to be disconnected over and above a downloading connection.

Reply:
yeah, that was a general comment. However, I think you can express this logic more straight-forward. For example, a peer that has been connected for less than 60 seconds could always be prioritized over one that has been connected longer. I think that could be done regardless of the number of bytes transferred.

Keep in mind that it can take a while for a TCP connection to ramp-up as well, especially if there's a lot of cross-traffic

Reply:
But that is not the logic. A connection <60s which is downloading slowly should NOT be prioritised over a connection that has been open longer but is faster.